SalesforceCommerceCloud/pwa-kit
# Support Node 16

Add support for node 16 by deleting `PerformanceTimer` class, and make node 16 the default version for new projects.

GUS: [W-12399737](https://gus.lightning.force.com/a07EE00001Jd7hjYAB)

# Description

Over on MRT, we're working on enabling platform-level support for node 16 apps. It currently doesn't work if the app depends on `pwa-kit-runtime` because of an issue with `PerformanceTimer`, our custom wrapper around the builtin node `perf_hooks` API. It turns out that we don't use `PerformanceTimer` for anything useful (it gathers metrics that we emit to CloudWatch, but Lambda automatically emits the same metrics to CloudWatch). Since the [class](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/develop/packages/pwa-kit-runtime/src/utils/ssr-server/performance-timer.js#L24) is marked as `@private`, this deletes it (which is not a breaking change) and adds explicit support for node 16.

# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes
- Remove `PerformanceTimer` and all its usages
- Explicitly support node 16 in all `package.json`s, and make `16.x` the default node version for new projects
- Update CircleCI workflow to include node 16 and npm 7 and 8
- Update `create-mobify-app-dev.js` to account for breaking changes in npm 7
- Fix (or make more resilient) a handful of `retail-react-app` tests, which were mostly failing due to race conditions that were uncovered with the switch to node 16, because the test execution environment became faster
- Re-introduce changes from #865
  - It was released in an alpha release (`pwa-kit-dev@2.4.1-alpha.0`), but somehow didn't make it into the next normal release, so running `npm run tail-logs` in a newly-generated project returns an error.

# How to Test-Drive This PR
- Run `npm ci` at the root of this repo
- Repeat the following for node 14.17.0 (npm 6), node 16.7.0 (npm 7), node 16.11.0 (npm 8), and node 16.19.1 (LTS, npm 8): 
  - Run `nvm use <node-version>`, then generate a new project:
  ```
  GENERATOR_PRESET=test-project node packages/pwa-kit-create-app/scripts/create-mobify-app-dev.js --outputDir 
  ~/generated-test-project
  ```
  - `cd` to the project, then run `npm start`. The app should load in the browser without errors.
  - Make sure your Runtime Admin API key is set in `~/.mobify`, then deploy your bundle to MRT: `npm run push -- -s <project-slug> -t <target-slug>`
    - In [Runtime Admin](https://runtime.commercecloud.com), make sure the deploy succeeds, that your site works, and that your bundle is running node `16.x`
    - For reference, I deployed my generated project bundle to https://zombie-apocalypse.mobify-storefront-staging.com
  - Run `npm run tail-logs -- -p scaffold-pwa -e production` and ensure logs appear after a few seconds.

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
